sql/rowexec/dml.go

```diff
@@ -54,7 +54,28 @@ func (b *BaseBuilder) buildInsertInto(ctx *sql.Context, ii *plan.InsertInto, row
 		return nil, err
 	}
 
+	var unlocker func()
 	insertExpressions := getInsertExpressions(ii.Source)
+	if ii.HasUnspecifiedAutoInc {
+		_, i, _ := sql.SystemVariables.GetGlobal("innodb_autoinc_lock_mode")
+		lockMode := i.(int64)
+		if err != nil {
+			return nil, err
+		}
+		// Lock modes "traditional" (0) and "consecutive" (1) require that a single lock is held for the entire iteration.
+		// Lock mode "interleaved" (2) will acquire the lock only when inserting into the table.
+		if lockMode != 2 {
+			autoIncrementable, ok := sql.GetUnderlyingTable(insertable).(sql.AutoIncrementTable)
+			if !ok {
+				panic("Auto increment expression on non-AutoIncrement table. This should not be possible.")
```

Comment:
Probably a little nicer to just return an `error` here instead of panic'ing. I know we have a panic handler in place, but returning an `error` seems like it accomplishes the same goal in a simpler way, allows any other error handling code to run, etc.